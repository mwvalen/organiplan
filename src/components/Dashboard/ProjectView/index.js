import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Redirect, withRouter } from 'react-router-dom'
import { deleteProject } from '../../../actions/manageProjects'

function msToTime(duration) {
    var milliseconds = parseInt((duration % 1000) / 100),
      seconds = parseInt((duration / 1000) % 60),
      minutes = parseInt((duration / (1000 * 60)) % 60),
      hours = parseInt((duration / (1000 * 60 * 60)) % 24);
  
    hours = (hours < 10) ? "0" + hours : hours;
    minutes = (minutes < 10) ? "0" + minutes : minutes;
    seconds = (seconds < 10) ? "0" + seconds : seconds;
  
    return hours + ":" + minutes + ":" + seconds + "." + milliseconds;
}

const DeleteProjectButton = ({project, deleteProject}) => {
    return (
        <div className='d-inline-block'>
        <button type="button" className="btn btn-outline-primary" data-toggle="modal" data-target="#deleteProjectModal" >
            Delete Project
        </button> 
        <div className="modal fade" id="deleteProjectModal" role="dialog" aria-labelledby="deleteNewProject" aria-hidden="true">
            <div className="modal-dialog" role="document">
                    <div className="modal-content">
                        <div className="modal-header">
                            <h5 className="modal-title">Delete Project</h5>
                            <button type="button" className="close" data-dismiss="modal" aria-label="Close">
                            <span aria-hidden="true">&times;</span>
                            </button>
                        </div>
                        <div className="modal-body">
                            <p>Are you sure you want to delete {project.name}?</p>
                        </div>
                        <div className="modal-footer">
                            <button type="button" className="btn btn-secondary" data-dismiss="modal" id="closeModal">Cancel</button>
                            <button type="button" className="btn btn-primary" onClick={deleteProject}>Yes, Delete</button>
                        </div>
                    </div>
                </div>
        </div>
        </div>
    )
}

export class ProjectView extends Component  {
    constructor(props) {
        super(props)
        this.state = {
            shouldRedirect: false
        }
    }
    deleteProject = () => {
        console.log('DELETING PROJECT')
        this.props.deleteProject(this.props.slug)
        var closeButton = document.getElementById('closeModal')
        closeButton.click()
        this.setState({...this.state, shouldRedirect: true })
    }

    render() {
        const project = this.props
        if (this.state.shouldRedirect) return <Redirect to="/" />
        return (
            <div className="col-md-10 pt-3">
                <div className="row pr-5 d-flex flex-row-reverse">
                    <DeleteProjectButton project={project} deleteProject={this.deleteProject.bind(this)} />
                </div>
                <div className="row pt-2 px-5">
                    <h3 className="my-5 w-100">{project.name}</h3>
                    <br /><br />
                    <div><h5 className="mb-3">Time Entries</h5></div>
                    <table className="table table-hover table-striped">
                    <thead>
                        <tr>
                        <th scope="col">#</th>
                        <th scope="col">Start</th>
                        <th scope="col">End</th>
                        <th scope="col">Total Time</th>
                        </tr>
                    </thead>
                    <tbody>
                    {project.timeEntries.map((te, i)=> 
                        <tr>
                            <th scope="row">{i+1}</th>
                            <td>{new Date(te.startTime*1000).toLocaleString()}</td>
                            <td>{new Date(te.endTime*1000).toLocaleString()}</td>
                            <td>{msToTime(((te.endTime - te.startTime)*1000))}</td>
                        </tr>
                    )
                    }
                    </tbody>
                    </table>
                </div>            
            </div>
        )
    }
}

const mapStateToProps = (state, ownProps)=>({
    ...state.projects.filter(p => p.slug === ownProps.slug)[0]
})

const mapDispatchToProps = (dispatch) => {
    return {
        deleteProject: (slug) => {
            dispatch(deleteProject(slug))
        }
    }
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(ProjectView))
