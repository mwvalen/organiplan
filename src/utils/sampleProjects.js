export default [
    {
        name: "Pack Snow",
        slug: "pack-snow",
        color: "800000",
        status: "active",
        timeEntries: [
            {
                startTime: "1545923460",
                endTime: "1545927240"
            },
            {
                startTime: "1546036800",
                endTime: "1546040400"
            },
            {
                startTime: "1546088400",
                endTime: "1546089470"
            },
        ]
    },
    {
        name: "Dig Tunnel",
        slug: "dig-tunnel",
        color: "FF0000",
        status: "active",
        timeEntries: [
            {
                startTime: "1545841070",
                endTime: "1545848252"
            },
            {
                startTime: "1545862652",
                endTime: "1545865500"
            },
            {
                startTime: "1545905100",
                endTime: "1545908700"
            },
        ]
    },
    {
        name: "Build Spaceship",
        slug: "build-spaceship",
        color: "FFA500",
        status: "active",
        timeEntries: [
            {
                startTime: "1545303900",
                endTime: "1545307500"
            },
            {
                startTime: "1545393900",
                endTime: "1545411900"
            },
            {
                startTime: "1545426300",
                endTime: "1545429900"
            },
        ]
    },
    {
        name: "Shovel Rocks",
        slug: "shovel-rocks",
        color: "008000",
        status: "active",
        timeEntries: [
            {
                startTime: "1545462300",
                endTime: "1545463020"
            },
            {
                startTime: "1545481020",
                endTime: "1545491820"
            },
            {
                startTime: "1545506220",
                endTime: "1545513420"
            },
            {
                startTime: "1545535020",
                endTime: "1545541500"
            },
            {
                startTime: "1545563100",
                endTime: "1545565954"
            },

        ]
    },
    {
        name: "Pilot Submarine",
        slug: "pilot-submarine",
        color: "808000",
        status: "active",
        timeEntries: [
        ]
    }
]