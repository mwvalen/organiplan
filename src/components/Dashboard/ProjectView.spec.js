import { mapStateToProps, DeleteProjectButton, TimeEntriesTable } from './ProjectView'
import renderer from 'react-test-renderer'
import React from 'react'
import sampleProjects from '../../utils/sampleProjects'

describe('The Project View Component',()=>{
    describe('The DeleteProjectButton',()=>{
        it('should not regress',()=>{
            const tree = renderer.create(
                <DeleteProjectButton
                    deleteProject={()=>{}}
                    project={sampleProjects[0]}
                />
            )

            expect(tree.toJSON()).toMatchSnapshot()
        })
    })
    describe('map state to props',()=>{
        it('should return the correct project',()=>{
            const state = {projects: sampleProjects}
            const props = mapStateToProps(state, {slug: sampleProjects[0].slug})

            expect(props).toEqual(sampleProjects[0])
        })
    })
    describe('the TimeEntriesTable',()=>{
        it('should not regress',()=>{
            const tree = renderer.create(
                <TimeEntriesTable project={sampleProjects[0]} />
            )

            expect(tree.toJSON()).toMatchSnapshot()
        })
    })
})