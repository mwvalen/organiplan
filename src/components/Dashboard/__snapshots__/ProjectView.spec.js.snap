// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Project View Component The DeleteProjectButton should not regress 1`] = `
<div
  className="d-inline-block"
>
  <button
    className="btn btn-outline-primary"
    data-target="#deleteProjectModal"
    data-toggle="modal"
    type="button"
  >
    Delete Project
  </button>
  <div
    aria-hidden="true"
    aria-labelledby="deleteNewProject"
    className="modal fade"
    id="deleteProjectModal"
    role="dialog"
  >
    <div
      className="modal-dialog"
      role="document"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <h5
            className="modal-title"
          >
            Delete Project
          </h5>
          <button
            aria-label="Close"
            className="close"
            data-dismiss="modal"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              Ã—
            </span>
          </button>
        </div>
        <div
          className="modal-body"
        >
          <p>
            Are you sure you want to delete 
            Pack Snow
            ?
          </p>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn btn-secondary"
            data-dismiss="modal"
            id="closeModal"
            type="button"
          >
            Cancel
          </button>
          <button
            className="btn btn-primary"
            onClick={[Function]}
            type="button"
          >
            Yes, Delete
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`The Project View Component the TimeEntriesTable should not regress 1`] = `
<table
  className="table table-hover table-striped"
>
  <thead>
    <tr>
      <th
        scope="col"
      >
        #
      </th>
      <th
        scope="col"
      >
        Start
      </th>
      <th
        scope="col"
      >
        End
      </th>
      <th
        scope="col"
      >
        Total Time
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th
        scope="row"
      >
        1
      </th>
      <td>
        2018-12-27 10:11:00
      </td>
      <td>
        2018-12-27 11:14:00
      </td>
      <td>
        01:03:00.0
      </td>
    </tr>
    <tr>
      <th
        scope="row"
      >
        2
      </th>
      <td>
        2018-12-28 17:40:00
      </td>
      <td>
        2018-12-28 18:40:00
      </td>
      <td>
        01:00:00.0
      </td>
    </tr>
    <tr>
      <th
        scope="row"
      >
        3
      </th>
      <td>
        2018-12-29 08:00:00
      </td>
      <td>
        2018-12-29 08:17:50
      </td>
      <td>
        00:17:50.0
      </td>
    </tr>
  </tbody>
</table>
`;
